### DISCUSSION TITLE
Go-to-Implementation and CodeLens for delegates

### DISCUSSION THREAD

**Discussion by glen-nicol** [dotnet/roslyn#58469] [Ideas] [answered]
This is largely a Visual studio feature request. But I think the related features are largely driven by Roslyn. I found [#23908](https://github.com/dotnet/roslyn/issues/23908) so assumed this would be related. If I am wrong I apologize. 

I have started using the excellent [LanguageExt](https://github.com/louthy/language-ext) library in my C# projects for awhile now and one thing I sorely miss about switching to delegates over interfaces is go-to-implementation. Specifically having a nice list of places where delegates are created or decorated/built up vs the potentially long list of references where the delegate is used. 

I realize that it is not as straight forward as finding the type in an inheritance list. But I think finding functions that return the delegate type explicitly would provide a good starting point and should be just as easy to implement. 

Other cases are not as clear:
* Properties/fields of the delegate type that are not simple reference assignments to existing. Meaning a location where a delegate is built up from other operators not just storing a reference for a DTO type object of delegates (poor mans interface I suppose). 
* Lambdas passed into a function as an argument for a parameter of the delegate type. Technically, an implementation but could be a lot of noise here too.

Finally, codelens on the delegate types would be helpful to more quickly see references to the delegate type at a glance.

**Comment by CyrusNajmabadi** [comment, marked as answer]
If you do a find references the results should have a column that allows you to group in this fashion.

**Reply by glen-nicol** [reply]
Thanks for the suggestion, I hadn't seen that column before. But at least on 2019 it doesn't have a category for return type. It only has one for "Type Argument" in my most used delegate which is helpful. But also it seems to be blank when the delegate is decorated. I assume because it is a type argument and also some other category (unknown/other default?) so it doesn't choose between them.

**Reply by CyrusNajmabadi** [reply]
hrmm... you mentioned:

> Specifically having a nice list of places where delegates are created

That can be done with find-refs like so:

![image](https://user-images.githubusercontent.com/4564579/147493256-0b093733-9234-4726-8132-37a30f941676.png)


Note how the second results says "constructor" as that's a reference to the delegate being created.

**Reply by glen-nicol** [reply]
Yes, but my delegates are typically created like this without constructors:
![image](https://user-images.githubusercontent.com/5103200/147493709-8edd4a77-05fc-4531-968e-f1a368962498.png)

I've also just noticed it returns multiple results for a single line. Without a kind I am not sure what the duplicate is for. I can think of these 
 possible categories: return Type, type argument,  invocation of delegate, delegate creation (the lambda)

**Reply by CyrusNajmabadi** [reply]
But those *aren't* creations.  Those are just literal references.  

> I've also just noticed it returns multiple results for a single line. Without a kind I am not sure what the duplicate is for. I can think of these

Each reference in the line gets a separate highlight.  You can see that here: 

![image](https://user-images.githubusercontent.com/4564579/147494116-9886ca93-2093-4fd5-afa2-be4240102c37.png)

> I can think of these possible categories

Sure. We'd take a focused PR that provided that.   :)

**Reply by glen-nicol** [reply]
Ahh thanks for pointing out the highlighting! I can do some searching to see about making a PR for those.

Okay, so are you concerned because a return type doesn't automatically mean delegate creation? I understand that won't always be the case and wasn't asking for a blanket assumption on the that alone. But could extra filtering for (return type + lambda) be used to infer delegate creation? I mean I suppose we could even just look at all lambda's to see if they match the delegate type. I figure that would give false positives though. Seems like the delegate proximity (as a return or as a parameter for which the lambda was created inline) would provide a good method for inferring.

**Reply by glen-nicol** [reply]
Seems like the groups that can be shown in VS are limited by the available enum values provided by visual studio in Microsoft.VisualStudio.LanguageServer.Protocol.VSInternalReferenceKind and conversions to it in [SymbolUsageInfoToReferenceKinds](https://github.com/dotnet/roslyn/blob/6884d36627795f1e00303bf4a068dd7998667bd9/src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs#L576). So Not sure there is much that can be done to improve those kind labels without a VS feature request for more language concepts to be added.

**Reply by CyrusNajmabadi** [reply]
> Okay, so are you concerned because a return type doesn't automatically mean delegate creation?

Correct.  They are distinct concepts.

**Reply by CyrusNajmabadi** [reply]
> So Not sure there is much that can be done to improve those kind labels without a VS feature request for more language concepts to be added.

Thats' waht this repo is for :)

**Reply by CyrusNajmabadi** [reply]
Note: delegate-creation-through-lambda is likely very hard for us to provide as there's no cheap way to know where all those references are going to show up.

**Reply by glen-nicol** [reply]
> > So Not sure there is much that can be done to improve those kind labels without a VS feature request for more language concepts to be added.
> 
> Thats' waht this repo is for :)

It is? That enum looked like it was part of the VS SDK. I didn't see that in the repo.

Kinda sounds like I need to write my own custom extension to do a search for these lambdas then? Since making it apart of find all references will likely be too expensive for normal usage?

**Reply by CyrusNajmabadi** [reply]
> It is? That enum looked like it was part of the VS SDK. I didn't see that in the repo.

This repo covers things like the VS find-refs/goto-impl features for C#/VB.  If changes to those experiences require changes in VS itself, we'll make that happen.  However, this repo is still the right place to create the issues/prs and coordinate things :)

> Kinda sounds like I need to write my own custom extension to do a search for these lambdas then? Since making it apart of find all references will likely be too expensive for normal usage?

This is quite possible.

**Reply by glen-nicol** [reply]
Happy New Year! How would I go about creating an issue for adding more find all reference categories? API proposal issue?

**Reply by CyrusNajmabadi** [reply]
You can just open a roslyn-issue.  Note that we should try to be very explicit about what these categories are, and precisely what sorts of cases they should appear for.

It should state if exsiting matches should get different categories.  Or if we need to find new matches altogether.  If we need to find new matches, we'll have to consider if that's even possible.

For new matches, again, please try to be crisp about exactly what should be matched that isn't being matched before.
